### PR TITLE
refactor(dom): consolidate transition/animation waiting utils

### PR DESCRIPTION
**Pull request by jcfranco**
**Related Issue:** N/A

## Summary

Stems from https://github.com/Esri/calcite-design-system/pull/9325#issue-2294070147.

Consolidates transition/animation waiting logic and adds `whenTransitionDone` to DOM utils to complement `whenAnimationDone`.

### CODE REVIEW

File: packages/calcite-components/src/utils/dom.ts [hunk 1]
```diff
@@ -661,27 +661,59 @@ export function isBefore(a: HTMLElement, b: HTMLElement): boolean {
...
-  const propIndex = allPropsArray.indexOf(animationName);
+  const allPropsArray = allProps.split(",");
+  const propIndex = allPropsArray.indexOf(transitionPropOrAnimationName);
   const duration =
     allDurationsArray[propIndex] ??
     /* Safari will have a single duration value for the shorthand prop when multiple, separate names/props are defined,
             so we fall back to it if there's no matching prop duration */
     allDurationsArray[0];
 
+  console.log("duration", duration, allDurations, allProps, style.transitionDuration, style.transitionProperty);
```
Comment (driskull) [inline packages/calcite-components/src/utils/dom.ts:705]: Should this be removed?

File: packages/calcite-components/src/utils/dom.spec.ts [hunk 2]
```diff
@@ -607,4 +610,140 @@ describe("dom", () => {
...
+
+  describe("whenTransitionDone", () => {
+    let dispatchTransitionEvent: (
+      element: HTMLElement,
+      type: "transitionstart" | "transitionend",
+      propertyName: string,
+    ) => void;
+
+    beforeEach(() => {
+      // we clobber Stencil's custom Mock document implementation
```
Comment (driskull) [inline packages/calcite-components/src/utils/dom.spec.ts:633]: 🔨

File: packages/calcite-components/src/utils/dom.spec.ts [hunk 3]
```diff
@@ -607,4 +610,140 @@ describe("dom", () => {
...
+
+    it("should return a promise that resolves after the transition", async () => {
+      const element = window.document.createElement("div");
+      const testProp = "opacity";
+      const testDuration = "0.5s";
+      const testTransition = `${testProp} ${testDuration} ease 0s`;
+
+      element.style.transition = testTransition;
+
+      // need to mock due to JSDOM issue with getComputedStyle - https://github.com/jsdom/jsdom/issues/3090
```
Comment (driskull) [inline packages/calcite-components/src/utils/dom.spec.ts:669]: 🎭

File: packages/calcite-components/src/utils/dom.ts [hunk 4]
```diff
@@ -695,19 +727,29 @@ export async function whenAnimationDone(targetEl: HTMLElement, animationName: st
     targetEl.addEventListener(endEvent, onEndOrCancel);
     targetEl.addEventListener(cancelEvent, onEndOrCancel);
 
-    function onStart(event: AnimationEvent): void {
-      if (event.animationName === animationName && event.target === targetEl) {
-        clearTimeout(fallbackTimeoutId);
+    function onStart(event: TransitionOrAnimationEvent): void {
+      console.log("start");
```
Comment (driskull) [inline packages/calcite-components/src/utils/dom.ts:731]: Moar console logs 🪵

File: packages/calcite-components/src/utils/dom.ts [hunk 5]
```diff
@@ -695,19 +727,29 @@ export async function whenAnimationDone(targetEl: HTMLElement, animationName: st
...
+      if (event.target === targetEl && getTransitionOrAnimationName(event) === transitionPropOrAnimationName) {
+        window.clearTimeout(fallbackTimeoutId);
         targetEl.removeEventListener(startEvent, onStart);
       }
     }
 
-    function onEndOrCancel(event: AnimationEvent): void {
-      if (event.animationName === animationName && event.target === targetEl) {
+    function onEndOrCancel(event: TransitionOrAnimationEvent): void {
+      console.log("end or c");
```
Comment (driskull) [inline packages/calcite-components/src/utils/dom.ts:739]: 🪵

File: packages/calcite-components/src/utils/dom.ts [hunk 6]
```diff
@@ -661,27 +661,59 @@ export function isBefore(a: HTMLElement, b: HTMLElement): boolean {
...
+  return whenTransitionOrAnimationDone(targetEl, animationName, "animation");
+}
+
+/**
+ * This util helps determine when a transition has completed.
+ *
+ * @param targetEl The element to watch for the transition to complete.
+ * @param transitionProp The name of the transition to watch for completion.
+ */
+export async function whenTransitionDone(targetEl: HTMLElement, transitionProp: string): Promise<void> {
```
Comment (driskull) [inline packages/calcite-components/src/utils/dom.ts:673]: Can't we just use `onTransitionEnd` on a VNode in most cases? Should we specify that?

File: patches/@types+jsdom+21.1.6.patch [hunk 7]
```diff
@@ -0,0 +1,14 @@
+diff --git a/node_modules/@types/jsdom/base.d.ts b/node_modules/@types/jsdom/base.d.ts
```
Comment (driskull) [inline patches/@types+jsdom+21.1.6.patch:1]: patches!

**Review by driskull — CHANGES_REQUESTED** [review]
Almost there!

File: patches/@types+jsdom+21.1.6.patch (hunk 7 above)
Comment (jcfranco) [reply patches/@types+jsdom+21.1.6.patch:1]: patches

File: packages/calcite-components/src/utils/dom.ts (hunk 4 above)
Comment (jcfranco) [reply packages/calcite-components/src/utils/dom.ts]: Yikes, sorry about that. All cleaned up now!

File: packages/calcite-components/src/utils/dom.ts (hunk 6 above)
Comment (jcfranco) [reply packages/calcite-components/src/utils/dom.ts:673]: I think we can tweak the name and doc to better explain its usage.

**Review by driskull — APPROVED** [review]
👍

**Pull request merged**